File: go/ppc/specification.go
package capitan

import (
	"regexp"
	"strings"
)

type Rule struct{}

type T struct {
	MatchTemplate   string
	Rule            *Rule
	RewriteTemplate *string
}

func Create(rewriteTemplate *string, rule *Rule, matchTemplate string) T {
	return T{
		MatchTemplate:   matchTemplate,
		Rule:            rule,
		RewriteTemplate: rewriteTemplate,
	}
}

type Extracted interface {
	toRegex() string
}

type Regex string
type ContiguousWhitespace string
type NonSpace string

func (r Regex) toRegex() string                 { return string(r) }
func (c ContiguousWhitespace) toRegex() string  { return "\\s+" }
func (n NonSpace) toRegex() string              { return regexp.QuoteMeta(string(n)) }

func escape(s string) string {
	return regexp.QuoteMeta(s)
}

func ToRegex(t T) string {
	extracted := extract(t.MatchTemplate)
	parts := make([]string, len(extracted))
	for i, e := range extracted {
		parts[i] = e.toRegex()
	}
	return "(" + strings.Join(parts, "") + ")"
}

func extract(template string) []Extracted {
	return nil
}

File: python/ppc/specification.py
import re
from typing import Optional, List, Union

class Rule:
    pass

class T:
    def __init__(self, match_template: str, rule: Optional[Rule], rewrite_template: Optional[str]):
        self.match_template = match_template
        self.rule = rule
        self.rewrite_template = rewrite_template

def create(rewrite_template: Optional[str], rule: Optional[Rule], match_template: str) -> T:
    return T(match_template, rule, rewrite_template)

class Extracted:
    def to_regex(self) -> str:
        raise NotImplementedError

class Regex(Extracted):
    def __init__(self, content: str):
        self.content = content

    def to_regex(self) -> str:
        return self.content

class ContiguousWhitespace(Extracted):
    def to_regex(self) -> str:
        return r'\s+'

class NonSpace(Extracted):
    def __init__(self, content: str):
        self.content = content

    def to_regex(self) -> str:
        return re.escape(self.content)

def extract(template: str) -> List[Extracted]:
    return []

def to_regex(t: T) -> str:
    extracted = extract(t.match_template)
    return '(' + ''.join(e.to_regex() for e in extracted) + ')'